src/main/java/com/cloudbees/plugins/credentials/CredentialsDescriptor.java

```diff
@@ -192,4 +257,340 @@ public String getIconClassName() {
+                                        // we have a winner
+                                        return type.cast(item);
+                                    }
+                                }
+                                // walk back up and try one level less deep
+                                pathSegments.remove(pathSegments.size() - 1);
+                            }
+                        }
+                    }
+                    // ok we give up, we are not thirsty for more, we'll let "normal" ancestor in path logic continue
+                }
+            }
+            if (type.isInstance(o) && o instanceof ModelObject && CredentialsProvider.hasStores((ModelObject) o)) {
+                return type.cast(o);
+            }
+        }
+        return null;
+
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public FormValidation.CheckMethod getCheckMethod(String fieldName) {
+        // this is an ugly hack to make the @ContextInPath annotation more failsafe
+        // requires that you explicitly call out the checkUrl: checkUrl="${descriptor.getCheckUrl('fieldName')}"
+        FormValidation.CheckMethod method = enhancedCheckMethods.get(fieldName);
+        if (method == null) {
+            method = new EnhancedCheckMethod(this, fieldName);
+            enhancedCheckMethods.put(fieldName, method);
+        }
+        return method;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void calcFillSettings(String field, Map<String, Object> attributes) {
+        if (attributes.containsKey("fillUrl")) {
+            // the user already provided a custom one, get out of the way
+            super.calcFillSettings(field, attributes);
+            return;
+        }
+        // this is an ugly hack to make the @ContextInPath annotation more failsafe
+        super.calcFillSettings(field, attributes);
+        if (attributes.containsKey("fillUrl")) {
+            try {
+                JellyContext jelly = Functions.getCurrentJellyContext();
+                Object it = jelly.findVariable("it");
+                if (it instanceof CredentialsStore) {
+                    ModelObject context = ((CredentialsStore) it).getContext();
+                    for (CredentialsSelectHelper.ContextResolver r : ExtensionList
+                            .lookup(CredentialsSelectHelper.ContextResolver.class)) {
+                        String token = r.getToken(context);
+                        if (token != null) {
+                            String fillUrl = (String) attributes.get("fillUrl");
+                            if (fillUrl.indexOf('?') != -1) {
+                                fillUrl = fillUrl + '&';
+                            } else {
+                                fillUrl = fillUrl + '?';
+                            }
+                            attributes.put("fillUrl", fillUrl + "$provider=" + r.getClass().getName() + "&"
```

Comment:
But it is safer to escape anyway... just pointing out that your stated reason is invalid... there is a valid reason to escape!